File: 2024/day13/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

function parseMachines(text) {
  const machines = [];
  const blocks = text.split('\n\n');

  for (const block of blocks) {
    const lines = block.trim().split('\n');
    // Button A: X+ax, Y+ay
    const aMatch = lines[0].match(/Button A: X\+(\d+), Y\+(\d+)/);
    const ax = parseInt(aMatch[1]), ay = parseInt(aMatch[2]);
    // Button B: X+bx, Y+by
    const bMatch = lines[1].match(/Button B: X\+(\d+), Y\+(\d+)/);
    const bx = parseInt(bMatch[1]), by = parseInt(bMatch[2]);
    // Prize: X=px, Y=py
    const pMatch = lines[2].match(/Prize: X=(\d+), Y=(\d+)/);
    const px = parseInt(pMatch[1]), py = parseInt(pMatch[2]);

    machines.push({ ax, ay, bx, by, px, py });
  }

  return machines;
}

function solveMachine(ax, ay, bx, by, px, py, maxPresses = null) {
  // Solve using Cramer's rule
  // a*ax + b*bx = px
  // a*ay + b*by = py
  //
  // det = ax*by - ay*bx
  // a = (px*by - py*bx) / det
  // b = (ax*py - ay*px) / det

  // Use BigInt for Part 2's large numbers
  const axB = BigInt(ax), ayB = BigInt(ay);
  const bxB = BigInt(bx), byB = BigInt(by);
  const pxB = BigInt(px), pyB = BigInt(py);

  const det = axB * byB - ayB * bxB;

  if (det === 0n) {
    return null; // No unique solution
  }

  const aNum = pxB * byB - pyB * bxB;
  const bNum = axB * pyB - ayB * pxB;

  // Check if solutions are integers
  if (aNum % det !== 0n || bNum % det !== 0n) {
    return null;
  }

  const a = aNum / det;
  const b = bNum / det;

  // Check non-negative
  if (a < 0n || b < 0n) {
    return null;
  }

  // Check max presses constraint (Part 1)
  if (maxPresses !== null && (a > BigInt(maxPresses) || b > BigInt(maxPresses))) {
    return null;
  }

  return 3n * a + b;
}

function part1() {
  const machines = parseMachines(input);
  let total = 0n;

  for (const { ax, ay, bx, by, px, py } of machines) {
    const cost = solveMachine(ax, ay, bx, by, px, py, 100);
    if (cost !== null) {
      total += cost;
    }
  }

  return total;
}

function part2() {
  const machines = parseMachines(input);
  const offset = 10_000_000_000_000;
  let total = 0n;

  for (const { ax, ay, bx, by, px, py } of machines) {
    const cost = solveMachine(ax, ay, bx, by, px + offset, py + offset, null);
    if (cost !== null) {
      total += cost;
    }
  }

  return total;
}

console.log('Part 1:', part1().toString());
console.log('Part 2:', part2().toString());

File: 2024/day13/go/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

type Machine struct {
	ax, ay int64
	bx, by int64
	px, py int64
}

func parseMachines(text string) []Machine {
	machines := []Machine{}
	blocks := strings.Split(strings.TrimSpace(text), "\n\n")

	buttonARegex := regexp.MustCompile(`Button A: X\+(\d+), Y\+(\d+)`)
	buttonBRegex := regexp.MustCompile(`Button B: X\+(\d+), Y\+(\d+)`)
	prizeRegex := regexp.MustCompile(`Prize: X=(\d+), Y=(\d+)`)

	for _, block := range blocks {
		lines := strings.Split(strings.TrimSpace(block), "\n")
		if len(lines) < 3 {
			continue
		}

		// Parse Button A
		aMatch := buttonARegex.FindStringSubmatch(lines[0])
		ax, _ := strconv.ParseInt(aMatch[1], 10, 64)
		ay, _ := strconv.ParseInt(aMatch[2], 10, 64)

		// Parse Button B
		bMatch := buttonBRegex.FindStringSubmatch(lines[1])
		bx, _ := strconv.ParseInt(bMatch[1], 10, 64)
		by, _ := strconv.ParseInt(bMatch[2], 10, 64)

		// Parse Prize
		pMatch := prizeRegex.FindStringSubmatch(lines[2])
		px, _ := strconv.ParseInt(pMatch[1], 10, 64)
		py, _ := strconv.ParseInt(pMatch[2], 10, 64)

		machines = append(machines, Machine{ax, ay, bx, by, px, py})
	}

	return machines
}

func solveMachine(ax, ay, bx, by, px, py int64, maxPresses *int64) *int64 {
	// Cramer's rule for system of linear equations:
	// a*ax + b*bx = px
	// a*ay + b*by = py
	//
	// det = ax*by - ay*bx
	// a = (px*by - py*bx) / det
	// b = (ax*py - ay*px) / det

	det := ax*by - ay*bx

	if det == 0 {
		return nil // No unique solution
	}

	// Calculate numerators
	aNum := px*by - py*bx
	bNum := ax*py - ay*px

	// Check if solutions are integers
	if aNum%det != 0 || bNum%det != 0 {
		return nil
	}

	a := aNum / det
	b := bNum / det

	// Check non-negative
	if a < 0 || b < 0 {
		return nil
	}

	// Check max presses constraint (Part 1)
	if maxPresses != nil && (a > *maxPresses || b > *maxPresses) {
		return nil
	}

	cost := 3*a + b
	return &cost
}

func part1(machines []Machine) int64 {
	total := int64(0)
	maxPresses := int64(100)

	for _, m := range machines {
		if cost := solveMachine(m.ax, m.ay, m.bx, m.by, m.px, m.py, &maxPresses); cost != nil {
			total += *cost
		}
	}

	return total
}

func part2(machines []Machine) int64 {
	total := int64(0)
	offset := int64(10_000_000_000_000)

	for _, m := range machines {
		// Shift prize coordinates
		if cost := solveMachine(m.ax, m.ay, m.bx, m.by, m.px+offset, m.py+offset, nil); cost != nil {
			total += *cost
		}
	}

	return total
}

func main() {
	// Read input file
	execPath, _ := os.Executable()
	execDir := filepath.Dir(execPath)
	inputPath := filepath.Join(execDir, "..", "input.txt")

	// Fallback to relative path if executable path doesn't work
	if _, err := os.Stat(inputPath); os.IsNotExist(err) {
		inputPath = "../input.txt"
	}

	data, err := os.ReadFile(inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input file: %v\n", err)
		os.Exit(1)
	}

	machines := parseMachines(string(data))

	fmt.Printf("Part 1: %d\n", part1(machines))
	fmt.Printf("Part 2: %d\n", part2(machines))
}
